# T_Matrix-Spectral_Analysis

In [5]:
!ls ./test_l3_l2_T/T

de_de_T.csv       en_zh-CN_T.csv    fr_ru_T.csv       ru_it_T.csv
de_en_T.csv       es_de_T.csv       fr_zh-CN_T.csv    ru_ru_T.csv
de_es_T.csv       es_en_T.csv       it_de_T.csv       ru_zh-CN_T.csv
de_fr_T.csv       es_es_T.csv       it_en_T.csv       zh-CN_de_T.csv
de_it_T.csv       es_fr_T.csv       it_es_T.csv       zh-CN_en_T.csv
de_ru_T.csv       es_it_T.csv       it_fr_T.csv       zh-CN_es_T.csv
de_zh-CN_T.csv    es_ru_T.csv       it_it_T.csv       zh-CN_fr_T.csv
en_de_T.csv       es_zh-CN_T.csv    it_ru_T.csv       zh-CN_it_T.csv
en_en_T.csv       fr_de_T.csv       it_zh-CN_T.csv    zh-CN_ru_T.csv
en_es_T.csv       fr_en_T.csv       ru_de_T.csv       zh-CN_zh-CN_T.csv
en_fr_T.csv       fr_es_T.csv       ru_en_T.csv
en_it_T.csv       fr_fr_T.csv       ru_es_T.csv
en_ru_T.csv       fr_it_T.csv       ru_fr_T.csv


In [1]:
import os
import pandas as pd
import numpy as np

### Set args


T_MATRIX_DIR = "./test_l3_l2_T/T"
VERBOSE = 3
#######

### Map args
T_matrix_dir = T_MATRIX_DIR
verbose = VERBOSE
#######

# create a list of items in T_matrix_dir
T_matrix_dir_items = os.listdir(T_matrix_dir)

# Loop through list of paths to process each matrix
T_matrix_dict = {}
for T_matrix_name in T_matrix_dir_items[:1]:
    
    # Create path name
    T_matrix_full_path = os.path.join(T_matrix_dir, T_matrix_name)
    
    # Extract language strings and extensions
    [T_matrix_lgs, T_matrix_ext] = T_matrix_name.split(".")
    T_matrix_lgs = T_matrix_lgs.rstrip("_T")
    
    # Filter for non csv extensions
    if T_matrix_ext != "csv":
        if verbose >= 1:
            print("Skipping %s item in %s: extenion is not '.csv'" % (T_matrix_name,T_matrix_dir))
    
    else:
        # Load in csv as numpy array
        T_matrix = np.loadtxt(open(T_matrix_full_path), delimiter=",")
        
        ### Add matrices to T_matrix_dict under Lg1-Lg-2 key
        T_matrix_dict[T_matrix_lgs] = {}
        
        # Raw matrix
        T_matrix_dict[T_matrix_lgs]["T_matrix"] = T_matrix
        
        # T_cov Covariance matrix TT^{T}
        T_cov = np.dot(T_matrix,T_matrix.T)
        T_matrix_dict[T_matrix_lgs]["T_cov"] = T_cov
        
        # T_inv Matrix Inverse T^{-1}
        
        # SVD U, \Sigma matrix, and V matrices

        ### Add language dict to full T_matrix_dict
         = lang_dict

        
        if verbose >= 2:
            print("Matrix %s loaded with shape %r." % (T_matrix_name, T_matrix.shape))
        
    
# print(T_matrix_lgs)
#     print(T_matrix_ext)
#     print(type(T_matrix_ext))
T_matrix_dict

Matrix de_de_T.csv loaded with shape (300, 300).


{'de_de': array([[  2.62798872e-02,  -8.00443720e-03,   9.06144828e-03, ...,
           4.51286905e-04,  -6.37024525e-04,  -6.43303106e-03],
        [ -7.71387713e-03,   3.17478180e-02,  -1.03924386e-02, ...,
          -3.13700526e-03,   1.11302815e-03,   9.44260973e-03],
        [  8.97303317e-03,  -1.00596156e-02,   3.05041317e-02, ...,
           5.88151684e-04,   6.44203828e-06,  -7.75917433e-03],
        ..., 
        [  8.34789593e-04,  -3.14916088e-03,   6.72287075e-04, ...,
           1.96522009e-02,  -1.65478152e-04,   1.01957284e-03],
        [  1.29190652e-04,   3.62422550e-04,   2.75136787e-04, ...,
           1.05115771e-03,   1.94369927e-02,   1.32160319e-03],
        [ -6.42255787e-03,   8.82536173e-03,  -7.61175994e-03, ...,
          -4.40414988e-05,   1.17071450e-03,   2.85430402e-02]])}

In [41]:
T_inv = np.linalg.inv(T_matrix)
T_inv

array([[ 94.89567226,  20.35967357,   6.7718968 , ...,   2.34332517,
         24.51701224,  -9.82782042],
       [ -2.96051441,  94.05096546,  20.97794425, ..., -15.18881932,
        -24.88619172,  80.05241732],
       [-22.65918501, -32.7518761 ,  53.05976914, ...,  -9.98179643,
        -21.08009261,   6.24493535],
       ..., 
       [ -8.57743751,  48.60726184,  29.88192204, ...,  86.94535171,
         -3.71001585,  72.91735552],
       [ 12.25899103,  -3.53288459,  16.69945066, ..., -19.36209148,
         78.07405285,  71.57359398],
       [-11.04576256, -23.58286134, -11.4602463 , ..., -18.75958708,
        -12.94199067,  90.87706923]])

In [44]:
T_by_T_inv = np.dot(T_matrix,T_inv)
T_by_T_inv.round(5)

array([[ 1.,  0.,  0., ..., -0.,  0., -0.],
       [-0.,  1.,  0., ...,  0.,  0., -0.],
       [-0.,  0.,  1., ..., -0., -0., -0.],
       ..., 
       [ 0.,  0., -0., ...,  1.,  0., -0.],
       [-0., -0., -0., ..., -0.,  1.,  0.],
       [-0.,  0.,  0., ..., -0.,  0.,  1.]])

In [38]:
np_answer = np.dot(T_matrix,T_tran)[0,:]
np_direct = [0,:]
dot_first_row = []
for i in range(300):
    print(np_answer[i], np_direct[i], np.dot(T_matrix[0,:],T_matrix[:,i]))
#     print(dot_first_row[i] - np_answer[i])

print(dot_first_row)

# print(T_matrix[0,:].tolist()[:10])
# print(T_tran[:,0].tolist()[:10])

0.0039021673627 0.0039021673627 0.0037940085339
-0.0040473810079 -0.0040473810079 -0.00399833613247
0.00376721735933 0.00376721735933 0.00374123077432
0.00498599789147 0.00498599789147 0.00487901977073
-0.00290273117037 -0.00290273117037 -0.00298954902859
-0.00462010470051 -0.00462010470051 -0.0045701257852
0.00315507558325 0.00315507558325 0.00312181978561
0.00151320939132 0.00151320939132 0.00148532907157
-0.00154726999838 -0.00154726999838 -0.00138808441207
-0.00270270669324 -0.00270270669324 -0.00261390635519
-0.000159819712855 -0.000159819712855 6.03619384976e-05
-0.000700569593592 -0.000700569593592 -0.000577766333886
-0.00173720331016 -0.00173720331016 -0.00171835028524
0.000719008720446 0.000719008720446 0.000735821365039
0.00145545696889 0.00145545696889 0.00148244058903
-0.00042235767253 -0.00042235767253 -0.000326039419624
0.00245874578166 0.00245874578166 0.00251467699738
-0.00235781305297 -0.00235781305297 -0.00233614585772
0.00147591203399 0.00147591203399 0.0014235143670

In [26]:
T_tran = T_matrix.T
T_cov = np.dot(T_matrix,T_tran)
print(np.dot(T_matrix[0],T_tran[0]))
display(np.dot(T_matrix,T_matrix))

display(T_matrix)
print(T_matrix[0].tolist()[:10])
print(T_tran[:][0].tolist()[:10])
# display(T_tran)
# display(T_cov)


0.0037940085339


array([[  3.79400853e-03,  -3.99833613e-03,   3.74123077e-03, ...,
          2.14881325e-04,  -4.00750555e-04,  -3.93402819e-03],
       [ -3.96567107e-03,   5.99035875e-03,  -4.74668324e-03, ...,
         -4.68861535e-04,   4.52156872e-04,   5.13800372e-03],
       [  3.67768717e-03,  -4.72474476e-03,   5.01318668e-03, ...,
          5.54791415e-05,  -4.36420889e-04,  -4.55180444e-03],
       ..., 
       [  1.56950907e-04,  -3.54034046e-04,   1.35169133e-05, ...,
          9.77759589e-04,   9.15033589e-05,   2.99635849e-06],
       [ -2.66717928e-04,   2.89106130e-04,  -2.99902717e-04, ...,
          1.52613504e-04,   7.74957173e-04,   4.36232387e-04],
       [ -3.82118356e-03,   5.00802397e-03,  -4.47773544e-03, ...,
         -1.40436752e-04,   5.36197672e-04,   5.72363250e-03]])

array([[  2.62798872e-02,  -8.00443720e-03,   9.06144828e-03, ...,
          4.51286905e-04,  -6.37024525e-04,  -6.43303106e-03],
       [ -7.71387713e-03,   3.17478180e-02,  -1.03924386e-02, ...,
         -3.13700526e-03,   1.11302815e-03,   9.44260973e-03],
       [  8.97303317e-03,  -1.00596156e-02,   3.05041317e-02, ...,
          5.88151684e-04,   6.44203828e-06,  -7.75917433e-03],
       ..., 
       [  8.34789593e-04,  -3.14916088e-03,   6.72287075e-04, ...,
          1.96522009e-02,  -1.65478152e-04,   1.01957284e-03],
       [  1.29190652e-04,   3.62422550e-04,   2.75136787e-04, ...,
          1.05115771e-03,   1.94369927e-02,   1.32160319e-03],
       [ -6.42255787e-03,   8.82536173e-03,  -7.61175994e-03, ...,
         -4.40414988e-05,   1.17071450e-03,   2.85430402e-02]])

[0.026279887184500694, -0.008004437200725079, 0.009061448276042938, 0.012618507258594036, -0.005415447521954775, -0.008686361834406853, 0.008724816143512726, 0.0036843381822109222, -0.003325438592582941, -0.005768531933426857]
[0.026279887184500694, -0.007713877130299807, 0.008973033167421818, 0.012088065035641193, -0.0052625639364123344, -0.009280218742787838, 0.009108894504606724, 0.003987328615039587, -0.0037526364903897047, -0.0068390341475605965]


In [68]:
?np.set_printoptions

In [4]:
from IPython.display import display, HTML

pd.options.display.max_rows = 90
pd.options.display.max_columns = 200

df = pd.read_csv(T_matrix_full_path, names=range(300))
df.head()

0         1         2         3         4         5         6    \
0  0.026280 -0.008004  0.009061  0.012619 -0.005415 -0.008686  0.008725   
1 -0.007714  0.031748 -0.010392 -0.012305  0.009108  0.011916 -0.007532   
2  0.008973 -0.010060  0.030504  0.014472 -0.006473 -0.010670  0.006090   
3  0.012088 -0.012271  0.015024  0.041155 -0.010766 -0.013844  0.011766   
4 -0.005263  0.008820 -0.005924 -0.010617  0.029115  0.009380 -0.007132   

        7         8         9         10        11        12        13   \
0  0.003684 -0.003325 -0.005769 -0.000817 -0.001629 -0.002036  0.000980   
1 -0.005184  0.004324  0.006496 -0.003059  0.002587  0.002850 -0.001647   
2  0.003015 -0.003181 -0.007588  0.001565 -0.002809 -0.003135  0.001605   
3  0.000982 -0.003505 -0.009180 -0.001073 -0.002930 -0.003866  0.004932   
4 -0.000059  0.001184  0.006593 -0.001456  0.001233  0.003861 -0.001215   

        14        15        16        17        18        19        20   \
0  0.002653 -0.000290  0.003938 -0.004112  0.003440 -0.001573 -0.003114   
1 -0.002947  0.002204 -0.004006  0.008060 -0.004319  0.002852  0.002112   
2  0.002630 -0.000859  0.005250 -0.006922  0.004210 -0.003514 -0.001346   
3  0.006947  0.001145  0.007360 -0.009088  0.004991 -0.004035 -0.004301   
4 -0.003837  0.000759 -0.004151  0.005421 -0.003345  0.002406  0.002346   

        21        22        23        24        25        26        27   \
0 -0.002909 -0.001654 -0.000874 -0.000060  0.005597 -0.000463 -0.001567   
1  0.005594 -0.000842  0.001690 -0.000260 -0.004397 -0.004014  0.000977   
2 -0.005710 -0.000394 -0.001425 -0.000699  0.004740  0.003743 -0.001963   
3 -0.005110  0.000406 -0.000841  0.000565  0.009509  0.000449 -0.003716   
4  0.005678 -0.000694  0.001849  0.000586 -0.005188 -0.001429  0.001655   

        28        29        30        31        32        33        34   \
0  0.006766  0.006126 -0.005909 -0.004805 -0.001347  0.005537  0.003756   
1 -0.008692 -0.007088  0.008842  0.004742 -0.000047 -0.006270 -0.007498   
2  0.008930  0.005329 -0.007845 -0.006219 -0.000499  0.003678  0.005896   
3  0.009401  0.005983 -0.007460 -0.007721 -0.001028  0.006212  0.006925   
4 -0.004837 -0.005424  0.004596  0.001937 -0.001686 -0.004305 -0.005190   

        35        36        37        38        39        40        41   \
0  0.002721  0.003895  0.002470  0.004184 -0.004590  0.000956  0.000143   
1 -0.003348 -0.003850 -0.001510 -0.004352  0.006044 -0.001624 -0.000057   
2  0.002589  0.002950  0.003697  0.000525 -0.004880  0.001664  0.004049   
3  0.002406  0.003235  0.005158  0.005773 -0.002839  0.003964  0.005501   
4 -0.002029  0.001040 -0.002130 -0.004201  0.001746 -0.001719 -0.000720   

        42        43        44        45        46        47        48   \
0  0.004943 -0.006687 -0.001861  0.005790 -0.002028  0.003358 -0.001408   
1 -0.007402  0.007801  0.001897 -0.008106  0.001119 -0.000440  0.002499   
2  0.004756 -0.008598 -0.002551  0.004827 -0.001493  0.001752 -0.001263   
3  0.005610 -0.012999 -0.002945  0.008515 -0.001330  0.000449 -0.004450   
4 -0.003328  0.004299  0.001203 -0.002872  0.000424 -0.000034  0.001370   

        49        50        51        52        53        54        55   \
0  0.004210 -0.003539 -0.001020 -0.000208 -0.002060 -0.000859  0.002053   
1 -0.003819  0.005058  0.001709  0.002922  0.002386  0.002296 -0.003759   
2  0.002125 -0.003466 -0.000688 -0.002165 -0.003558 -0.001682  0.001880   
3  0.006653 -0.004091  0.000208 -0.004035 -0.003320 -0.002771  0.000396   
4 -0.003103  0.002731  0.000885  0.003639  0.001346  0.002668 -0.002734   

        56        57        58        59        60        61        62   \
0  0.004026  0.001058  0.000715  0.000500  0.002651 -0.001065 -0.005070   
1 -0.005017 -0.003030 -0.000109 -0.000054 -0.004471 -0.000454  0.005101   
2  0.004827  0.003003  0.000511 -0.000904  0.002554 -0.003325 -0.005340   
3  0.004606  0.005559  0.001181 -0.001142  0.003037 -0.000690 -0.005746   
4 -0.002068 -0.004790 -0.

In [43]:
df.shape

(300, 300)

In [37]:
?pd.read_csv

In [23]:
T_matrix_full_path

'./test_l3_l2_T/T/de_de_T.csv'

In [ ]:
# Set list of T_matrix full paths
T_matrix_dir = "./test_l3_l2_T/T"
T_matrix_names = !ls $T_matrix_dir

T_matrix_full_paths = []
for T_matrix_name in T_matrix_names:
    T_matrix_loc = T_matrix_dir+T_matrix_name
    T_matrix_full_paths += [T_matrix_loc]

# Load T_matrix into memory
load_path = T_matrix_full_paths[2]
print(load_path)
file = open ("./T_Matrices_examples/T_matrix_de_es.pkl",'rb')
file.seek(0)
object_file = pickle.load(file)

"""
with open(r"./T_Matrices_examples/T_matrix_de_es.pkl", 'rb') as f:
    T_matrix = pickle.load(f)"""

In [ ]:
     U,s,Vh = SVD(T)
        
        s1 = log(s)
    
        
        for stat in stats:
            svd[stat,translation[1],translation[2]] = stat_calc(stat, s, fro, acc)
            svd1[stat,translation[1],translation[2]] = stat_calc(stat, s1, fro, acc)
            
        
    #Exporting DataFrames for SVD Heatmaps
    
    s_df = make_df(languages,languages)
    s1_df = make_df(languages,languages)
    

    for stat in stats:
        for lang1 in languages:
            for lang2 in languages:
                s_df.set_value(lang1,lang2,svd[stat,lang1,lang2])
                s1_df.set_value(lang1,lang2,svd1[stat,lang1,lang2])

        s_df.to_csv('../HeatmapData/T/s_{}.csv'.format(stat),columns = languages)
        s1_df.to_csv('../HeatmapData/T/s1_{}.csv'.format(stat),columns = languages)


In [ ]:
***** NOTE THE BELOW HAS BEEN MOVED TO DRIVER NOTEBOOK CURRENTLY IN CODE *******

## Overview
The Isomantics algorithm consists of the following stages:
### **(Stage 0)** Prepare vocabulary
* currently prepared languages:
    1. English
    2. Russian
    3. German
    4. French
    5. Italian
    6. Chinese 
    
### **(Stage 1)** Word embeddings.  Using `fasttext` or `word2vec` we embed the vocab for each of the languages
### **(Stage 2)** Train translation matrices:

* Training set:
    * For two given languages $Lg_1$ and $Lg_2$, we create a training set $\Omega_{(Lg_1,Lg_2)}$ as follows:
        1. For each $word_i$ in language 1, find the direct translation $\widehat{word_i}$ in language 2.
        2. Find vector embeddings $w_i\in Lg_1$ and $\widehat{w_i}\in Lg_2$ of $word_i$ and $\widehat{word_i}$ respectively.
        3. Add the pair $<w_i,\widehat{w_i}>$ to the training set $\Omega_{(Lg_1,Lg_2)}$
            * **Note** we found that training only for for only the the top 5-10k most popular terms in  $\Omega_{(Lg_1,Lg_2)}$ generates the best word-to-word translation results on out of sample test sets.
* Building the Cost function:
    * Loss function for the learning process:
        * $ Loss(T_{Lg_1,Lg_2})= ||Tw_i - \widehat{w_i}||^2_2 $
    * Regularization terms:
        * Over fitting Regularizer:
            * $Reg_{Frobenius}(T_{Lg_1,Lg_2}) = ||T_{Lg_1,Lg_2}||_2$
        * Normality Regularizer:
            * $Reg_{Normality}(T_{Lg_1,Lg_2}) = ||T_{Lg_1,Lg_2}^{T}T_{Lg_1,Lg_2} - T_{Lg_1,Lg_2}T_{Lg_1,Lg_2}^T||_2$
                * **Note** The Normality Regularizer is used to ensure that the resulting matrix is diagonalizable.


#### Full cost function:
$$ J(T_{Lg_1,Lg_2})= Loss(T_{Lg_1,Lg_2}) + \lambda_{1}Reg_{Frobenius}(T_{Lg_1,Lg_2}) + \lambda_{2}Reg_{Normality}(T_{Lg_1,Lg_2}) $$  

### **(Stage 3)** Translation Spectral Analysis:
* Factor the matrix $T_{Lg_1,Lg_2} = U\Sigma V^T$ where $U$ and $V$ are orthonormal (rotation) matrices and $\Sigma$ gives the eigenvalues of $T_{Lg_1,Lg_2}$  or the "*Translation spectrum*"
* Run a statistical analysis of the spectral values associated with each pair of languages.
    1. mean
    2. median
    3. max value
    4. min value
    5. standard deviation

* Compare the statistical spectral analysis across different language pairs.

# (Stage 0) Prepare vocabulary:

# TODO 
 * add details on where the vocab was downloaded from
 * point to where the data is in the repo
 * add instructions on where to call it in the code

# (Stages 1 and 2) 
1. Word embeddings. Using fasttext or word2vec we embed the vocab for each of the languages
2. Train translation matrices:

# TODO 
 * add details on where the vocab is located
 * point to where the embeddings are located
 * Test following code on embedding process:

In [1]:
# Import tools for running Spectral decomposition
import sys
sys.path.append("../")
import ismtools



Using TensorFlow backend.


In [ ]:
# TODO script this section and call the script from bash in this cell
    # add the -a for setting calculate_KNN = True
# TODO?? put manual list of translations into ismtools or another imported .py

#Set parameter on calculate KNN (need to change for -a sysarg)
calculate_KNN = False

languages = ['en','ru','de','es','fr','it', 'zh-CN']
for lang1 in languages:
    for lang2 in languages:
        translations.append(('fasttext_top',lang1, lang2))

for translation in translations:
    embedding, lg1, lg2 = translation
    # Vocab/Vectors/Dicts
    lg1_vocab, lg1_vectors, lg2_vocab, lg2_vectors = \
        pickle_rw((lg1 + '_' + embedding.split('_')[0] + '_vocab', 0),
                  (lg1 + '_' + embedding.split('_')[0] + '_vectors', 0),
                  (lg2 + '_' + embedding.split('_')[0] + '_vocab', 0),
                  (lg2 + '_' + embedding.split('_')[0] + '_vectors', 0),
                  write=False)
    lg1_dict = make_dict(lg1_vocab, lg1_vectors)
    lg2_dict = make_dict(lg2_vocab, lg2_vectors)

    print('Translation: '+lg1+'->'+lg2+'\n')

    # Train/Test Vocab/Vectors
    vocab_train, vocab_test = vocab_train_test(embedding, lg1, lg2, lg1_vocab)
    X_train, X_test, y_train, y_test = vectors_train_test(vocab_train,
                                                          vocab_test,lg1_dict,lg2_dict)
    
    # Fit tranlation matrix to training data
    model, history, T, tf, I, M, fro = translation_matrix(X_train, y_train)
    
    if calculate_KNN:
        results_df = translation_results(X_test, y_test, vocab_test, T,
                                     lg2_vectors, lg2_vocab)
        acc = T_norm_EDA(results_df)

"""
TODO create standardized dumping location for the translation matrix and its meta data.
- pickel dump based on the dir structure in the folders
"""        

# (Stage 3) Translation Spectral Analysis

# TODO 
 * 

In [42]:
# Set list of T_matrix full paths
T_matrix_dir = "./T_Matrices_examples/"
T_matrix_names = !ls $T_matrix_dir

T_matrix_full_paths = []
for T_matrix_name in T_matrix_names:
    T_matrix_loc = T_matrix_dir+T_matrix_name
    T_matrix_full_paths += [T_matrix_loc]

# Load T_matrix into memory
load_path = T_matrix_full_paths[2]
print(load_path)
file = open ("./T_Matrices_examples/T_matrix_de_es.pkl",'rb')
file.seek(0)
object_file = pickle.load(file)

"""
with open(r"./T_Matrices_examples/T_matrix_de_es.pkl", 'rb') as f:
    T_matrix = pickle.load(f)"""

./T_Matrices_examples/T_matrix_de_es.pkl


EOFError: 

In [ ]:
U,s,Vh = SVD(T)

s1 = log(s)


for stat in stats:
    svd[stat,translation[1],translation[2]] = stat_calc(stat, s, fro, acc)
    svd1[stat,translation[1],translation[2]] = stat_calc(stat, s1, fro, acc)


#Exporting DataFrames for SVD Heatmaps

s_df = make_df(languages,languages)
s1_df = make_df(languages,languages)


for stat in stats:
    for lang1 in languages:
        for lang2 in languages:
            s_df.set_value(lang1,lang2,svd[stat,lang1,lang2])
            s1_df.set_value(lang1,lang2,svd1[stat,lang1,lang2])

    s_df.to_csv('../HeatmapData/T/s_{}.csv'.format(stat),columns = languages)
    s1_df.to_csv('../HeatmapData/T/s1_{}.csv'.format(stat),columns = languages)



In [ ]:
if __name__ == '__main__':
    # Manually set list of translations (embedding, lg1, lg2)
    
    svds = ['s','s1']
    languages = ['en','ru','de','es','fr','it', 'zh-CN']
    stats = ['min','max','mean','median','std','fro','acc']
    
    translations=[]
    
    for lang1 in languages:
        for lang2 in languages:
            translations.append(('fasttext_top',lang1, lang2))
    
    svd = {}
    svd1 = {}
    
    for translation in translations:
        embedding, lg1, lg2 = translation
        # Vocab/Vectors/Dicts
        lg1_vocab, lg1_vectors, lg2_vocab, lg2_vectors = \
            pickle_rw((lg1 + '_' + embedding.split('_')[0] + '_vocab', 0),
                      (lg1 + '_' + embedding.split('_')[0] + '_vectors', 0),
                      (lg2 + '_' + embedding.split('_')[0] + '_vocab', 0),
                      (lg2 + '_' + embedding.split('_')[0] + '_vectors', 0),
                      write=False)
        lg1_dict = make_dict(lg1_vocab, lg1_vectors)
        lg2_dict = make_dict(lg2_vocab, lg2_vectors)

        print('Translation: '+lg1+'->'+lg2+'\n')

        # Train/Test Vocab/Vectors
        vocab_train, vocab_test = vocab_train_test(embedding, lg1, lg2, lg1_vocab)
        X_train, X_test, y_train, y_test = vectors_train_test(vocab_train,
                                                              vocab_test,lg1_dict,lg2_dict)
 
        
        # Fit tranlation matrix to training data
        model, history, T, tf,I, M, fro = translation_matrix(X_train, y_train)
        
        results_df = translation_results(X_test, y_test, vocab_test, T,
                                         lg2_vectors, lg2_vocab)
        acc = T_norm_EDA(results_df)
        
        U,s,Vh = SVD(T)
        
        s1 = log(s)
    
        
        for stat in stats:
            svd[stat,translation[1],translation[2]] = stat_calc(stat, s, fro, acc)
            svd1[stat,translation[1],translation[2]] = stat_calc(stat, s1, fro, acc)
            
        
    #Exporting DataFrames for SVD Heatmaps
    
    s_df = make_df(languages,languages)
    s1_df = make_df(languages,languages)
    

    for stat in stats:
        for lang1 in languages:
            for lang2 in languages:
                s_df.set_value(lang1,lang2,svd[stat,lang1,lang2])
                s1_df.set_value(lang1,lang2,svd1[stat,lang1,lang2])

        s_df.to_csv('../HeatmapData/T/s_{}.csv'.format(stat),columns = languages)
        s1_df.to_csv('../HeatmapData/T/s1_{}.csv'.format(stat),columns = languages)

    



# Translation Matrix Results  
## En to Ru Fasttext_Random  
- En Vocabulary Size = 1,259,685  
- En Embedding Length = 300  
- Ru Vocabulary Size = 944,211  
- Ru Embedding Length = 300  
- Train Size = 5,000  
- Test Size = 1,500  
- <b>Test Accuracy = 3.9%</b>  

#### Test L2 Norms  
- X_norm: L2 norms for En test vectors  
- y_norm: L2 norms for Ru test vectors  
- yhat_norm: L2 norms for X.dot(T) test vectors (T = translation matrix)  
- yhat_neighbor norm: L2 norms for nearest neighborto X.dot(T) in y test vectors  
![](../images/en_ru_fasttext_random_T_norm.png)  

#### Translation Matrix Isotropy  
- Isotropy = 32.3%  
![](../images/en_ru_fasttext_random_T_isotropy.png)  

## En to Ru Fasttext_Top  
- En Vocabulary Size = 1,259,685  
- En Embedding Length = 300  
- Ru Vocabulary Size = 944,211  
- Ru Embedding Length = 300  
- Train Size = 5,000  
- Test Size = 1,500  
- <b>Test Accuracy = 46.3%</b>  

#### Test L2 Norms  
- X_norm: L2 norms for En test vectors  
- y_norm: L2 norms for Ru test vectors  
- yhat_norm: L2 norms for X.dot(T) test vectors (T = translation matrix)  
- yhat_neighbor norm: L2 norms for nearest neighborto X.dot(T) in y test vectors  
![](../images/en_ru_fasttext_top_T_norm.png)  

#### Translation Matrix Isotropy  
- Isotropy = 38.2%  
![](../images/en_ru_fasttext_top_T_isotropy.png)  

## En to De Fasttext_Random  
- En Vocabulary Size = 1,259,685  
- En Embedding Length = 300  
- De Vocabulary Size = 1,137,616  
- De Embedding Length = 300  
- Train Size = 5,000  
- Test Size = 1,500  
- <b>Test Accuracy = 21.9%</b>  

#### Test L2 Norms  
- X_norm: L2 norms for En test vectors  
- y_norm: L2 norms for De test vectors  
- yhat_norm: L2 norms for X.dot(T) test vectors (T = translation matrix)  
- yhat_neighbor norm: L2 norms for nearest neighborto X.dot(T) in y test vectors  
![](../images/en_de_fasttext_random_T_norm.png)  

#### Translation Matrix Isotropy  
- Isotropy = 35.6%  
![](../images/en_de_fasttext_random_T_isotropy.png)  

## En to De Fasttext_Top  
- En Vocabulary Size = 1,259,685  
- En Embedding Length = 300  
- De Vocabulary Size = 1,137,616  
- De Embedding Length = 300  
- Train Size = 5,000  
- Test Size = 1,500  
- <b>Test Accuracy = 63.6%</b>  

#### Test L2 Norms  
- X_norm: L2 norms for En test vectors  
- y_norm: L2 norms for De test vectors  
- yhat_norm: L2 norms for X.dot(T) test vectors (T = translation matrix)  
- yhat_neighbor norm: L2 norms for nearest neighborto X.dot(T) in y test vectors  
![](../images/en_de_fasttext_top_T_norm.png)  

#### Translation Matrix Isotropy  
- Isotropy = 43.4%  
![](../images/en_de_fasttext_top_T_isotropy.png)  

## En to It Zeroshot  
- En Vocabulary Size = 200,000  
- En Embedding Length = 300  
- It Vocabulary Size = 200,000  
- It Embedding Length = 300  
- Train Size = 5,000  
- Test Size = 1,869  
- <b>Test Accuracy = 27.9%</b>  

#### Test L2 Norms  
- X_norm: L2 norms for En test vectors  
- y_norm: L2 norms for It test vectors  
- yhat_norm: L2 norms for X.dot(T) test vectors (T = translation matrix)  
- yhat_neighbor norm: L2 norms for nearest neighborto X.dot(T) in y test vectors  
![](../images/en_it_zeroshot_T_norm.png)  

#### Translation Matrix Isotropy  
- Isotropy = 46.6%  
![](../images/en_it_zeroshot_T_isotropy.png)  

